File: ui/lib/api/attachments.ts
/**
 * 版权所有 (c) 2023-2026 北京慧测信息技术有限公司(但问智能) 保留所有权利。
 * 
 * 本代码版权归北京慧测信息技术有限公司(但问智能)所有，仅用于学习交流目的，未经公司商业授权，
 * 不得用于任何商业用途，包括但不限于商业环境部署、售卖或以任何形式进行商业获利。违者必究。
 * 
 * 授权商业应用请联系微信：huice666
 */

/**
 * 附件 API
 */
// FIXME  MC80OmFIVnBZMlhsdktEbHVwNDZUM0pyYWc9PTo2ZDk1ZDllZA==

import { apiClient } from "./client";

export interface AttachmentInfo {
  id: string;
  name: string;
  size: number;
  content_type?: string;
  created_by?: string;
  created_at: string;
  url?: string;
}

export interface AttachmentUploadResponse {
  id: string;
  name: string;
  size: number;
  content_type?: string;
  created_by?: string;
  created_at: string;
  url?: string;
}
// FIXME  MS80OmFIVnBZMlhsdktEbHVwNDZUM0pyYWc9PTo2ZDk1ZDllZA==

/**
 * 上传测试用例附件
 */
export async function uploadTestCaseAttachment(
  projectIdentifier: string,
  testCaseIdentifier: string,
  file: File
): Promise<AttachmentUploadResponse> {
  const formData = new FormData();
  formData.append("file", file);

  // 直接使用 fetch，因为 apiClient 会将 body 转换为 JSON
  const response = await fetch(
    `/api/v2/projects/${projectIdentifier}/test-cases/${testCaseIdentifier}/attachments`,
    {
      method: "POST",
      body: formData,
    }
  );

  if (!response.ok) {
    throw new Error(`Upload failed: ${response.statusText}`);
  }

  return response.json();
}

/**
 * 获取测试用例附件列表
 */
export async function getTestCaseAttachments(
  projectIdentifier: string,
  testCaseIdentifier: string
): Promise<AttachmentInfo[]> {
  const response = await apiClient.get<{ attachments: AttachmentInfo[] }>(
    `/api/v2/projects/${projectIdentifier}/test-cases/${testCaseIdentifier}/attachments`
  );

  return response.attachments;
}
// FIXME  Mi80OmFIVnBZMlhsdktEbHVwNDZUM0pyYWc9PTo2ZDk1ZDllZA==

/**
 * 删除测试用例附件
 */
export async function deleteTestCaseAttachment(
  projectIdentifier: string,
  testCaseIdentifier: string,
  attachmentId: string
): Promise<void> {
  await apiClient.delete(
    `/api/v2/projects/${projectIdentifier}/test-cases/${testCaseIdentifier}/attachments/${attachmentId}`
  );
}

/**
 * 获取附件下载链接
 */
export async function getAttachmentDownloadUrl(
  projectIdentifier: string,
  attachmentId: string
): Promise<string> {
  const response = await apiClient.get<{ url: string }>(
    `/api/v2/projects/${projectIdentifier}/attachments/${attachmentId}`
  );

  return response.url;
}
// FIXME  My80OmFIVnBZMlhsdktEbHVwNDZUM0pyYWc9PTo2ZDk1ZDllZA==

File: backend/app/api/v2/attachments.py
"""
附件管理 API

提供附件的上传、下载和删除操作
参考: https://www.browserstack.com/docs/test-management/api-reference/attachments
"""
"""
版权所有 (c) 2023-2026 北京慧测信息技术有限公司(但问智能) 保留所有权利。

本代码版权归北京慧测信息技术有限公司(但问智能)所有，仅用于学习交流目的，未经公司商业授权，
不得用于任何商业用途，包括但不限于商业环境部署、售卖或以任何形式进行商业获利。违者必究。

授权商业应用请联系微信：huice666
"""


from typing import Annotated
from uuid import UUID

from fastapi import APIRouter, status, UploadFile, File, Depends, Response
from fastapi.responses import StreamingResponse
from sqlalchemy.ext.asyncio import AsyncSession
import io

from app.api.deps import DbSessionDep
from app.services.attachment_service import AttachmentService
from app.schemas.attachment import (
    AttachmentUploadResponse,
    TestCaseAttachmentsResponse,
    TestResultAttachmentsResponse,
)
from app.schemas.common import SuccessResponse, MessageResponse
from app.config.database import get_db


# ============ 测试用例附件路由 ============
test_case_attachments_router = APIRouter(
    prefix="/projects/{project_identifier}/test-cases/{test_case_identifier}/attachments"
)
# noqa  MC80OmFIVnBZMlhsdktEbHVwNDZUakl4U1E9PTo4YmUxYTBiMA==


def get_attachment_service(session: AsyncSession = Depends(get_db)) -> AttachmentService:
    """获取附件服务"""
    return AttachmentService(session)

# pragma: no cover  MS80OmFIVnBZMlhsdktEbHVwNDZUakl4U1E9PTo4YmUxYTBiMA==

AttachmentServiceDep = Annotated[AttachmentService, Depends(get_attachment_service)]


@test_case_attachments_router.get(
    "",
    response_model=SuccessResponse[TestCaseAttachmentsResponse],
    summary="获取测试用例附件列表",
    description="获取指定测试用例的所有附件",
)
async def get_test_case_attachments(
    project_identifier: str,
    test_case_identifier: str,
    service: AttachmentServiceDep,
) -> SuccessResponse[TestCaseAttachmentsResponse]:
    """获取测试用例附件列表"""
    result = await service.get_test_case_attachments(
        project_identifier, test_case_identifier
    )
    return SuccessResponse(success=True, data=result)


@test_case_attachments_router.post(
    "",
    response_model=SuccessResponse[AttachmentUploadResponse],
    status_code=status.HTTP_201_CREATED,
    summary="上传测试用例附件",
    description="为测试用例上传附件文件，最大支持 50MB",
)
async def upload_test_case_attachment(
    project_identifier: str,
    test_case_identifier: str,
    file: UploadFile = File(..., description="要上传的文件"),
    service: AttachmentServiceDep = None,
    db: DbSessionDep = None,
) -> SuccessResponse[AttachmentUploadResponse]:
    """上传测试用例附件"""
    result = await service.upload_test_case_attachment(
        project_identifier, test_case_identifier, file
    )
    await db.commit()
    return SuccessResponse(success=True, data=result)


@test_case_attachments_router.delete(
    "/{attachment_id}",
    response_model=MessageResponse,
    summary="删除测试用例附件",
    description="删除指定的测试用例附件，此操作不可撤销",
)
async def delete_test_case_attachment(
    project_identifier: str,
    test_case_identifier: str,
    attachment_id: UUID,
    service: AttachmentServiceDep,
    db: DbSessionDep,
) -> MessageResponse:
    """删除测试用例附件"""
    await service.delete_test_case_attachment(
        project_identifier, test_case_identifier, str(attachment_id)
    )
    await db.commit()
    return MessageResponse(success=True, message="附件删除成功")


# ============ 测试结果附件路由 ============
test_result_attachments_router = APIRouter(
    prefix="/projects/{project_identifier}/test-results/{test_result_id}/attachments"
)


@test_result_attachments_router.get(
    "",
    response_model=SuccessResponse[TestResultAttachmentsResponse],
    summary="获取测试结果附件列表",
    description="获取指定测试结果的所有附件",
)
async def get_test_result_attachments(
    project_identifier: str,
    test_result_id: UUID,
    service: AttachmentServiceDep,
) -> SuccessResponse[TestResultAttachmentsResponse]:
    """获取测试结果附件列表"""
    result = await service.get_test_result_attachments(
        project_identifier, str(test_result_id)
    )
    return SuccessResponse(success=True, data=result)
# pylint: disable  Mi80OmFIVnBZMlhsdktEbHVwNDZUakl4U1E9PTo4YmUxYTBiMA==


@test_result_attachments_router.post(
    "",
    response_model=SuccessResponse[AttachmentUploadResponse],
    status_code=status.HTTP_201_CREATED,
    summary="上传测试结果附件",
    description="为测试结果上传附件文件，最大支持 50MB",
)
async def upload_test_result_attachment(
    project_identifier: str,
    test_result_id: UUID,
    file: UploadFile = File(..., description="要上传的文件"),
    service: AttachmentServiceDep = None,
    db: DbSessionDep = None,
) -> SuccessResponse[AttachmentUploadResponse]:
    """上传测试结果附件"""
    result = await service.upload_test_result_attachment(
        project_identifier, str(test_result_id), file
    )
    await db.commit()
    return SuccessResponse(success=True, data=result)


@test_result_attachments_router.delete(
    "/{attachment_id}",
    response_model=MessageResponse,
    summary="删除测试结果附件",
    description="删除指定的测试结果附件，此操作不可撤销",
)
async def delete_test_result_attachment(
    project_identifier: str,
    test_result_id: UUID,
    attachment_id: UUID,
    service: AttachmentServiceDep,
    db: DbSessionDep,
) -> MessageResponse:
    """删除测试结果附件"""
    await service.delete_test_result_attachment(
        project_identifier, str(test_result_id), str(attachment_id)
    )
    await db.commit()
    return MessageResponse(success=True, message="附件删除成功")

# noqa  My80OmFIVnBZMlhsdktEbHVwNDZUakl4U1E9PTo4YmUxYTBiMA==

# ============ 通用附件路由 ============
attachments_router = APIRouter(
    prefix="/projects/{project_identifier}/attachments"
)


@attachments_router.get(
    "/{attachment_id}",
    summary="获取附件下载链接",
    description="获取附件的预签名下载 URL",
)
async def get_attachment_download_url(
    project_identifier: str,
    attachment_id: UUID,
    service: AttachmentServiceDep,
):
    """获取附件下载链接"""
    result = await service.get_attachment_download_url(
        project_identifier, str(attachment_id)
    )
    return SuccessResponse(success=True, data=result)


@attachments_router.get(
    "/{attachment_id}/download",
    summary="下载附件",
    description="直接下载附件文件",
    responses={
        200: {
            "content": {"application/octet-stream": {}},
            "description": "附件文件内容",
        }
    },
)
async def download_attachment(
    project_identifier: str,
    attachment_id: UUID,
    service: AttachmentServiceDep,
) -> StreamingResponse:
    """下载附件"""
    content, filename, content_type = await service.download_attachment(
        project_identifier, str(attachment_id)
    )

    return StreamingResponse(
        io.BytesIO(content),
        media_type=content_type,
        headers={
            "Content-Disposition": f'attachment; filename="{filename}"',
            "Content-Length": str(len(content)),
        }
    )



# ============ 全局附件路由（按 ID 直接访问，不需要 project） ============
global_attachments_router = APIRouter(
    prefix="/attachments"
)


@global_attachments_router.get(
    "/{attachment_id}/content",
    summary="按附件 ID 下载附件内容",
    description="直接通过附件 ID 下载附件文件内容（无需 project 前缀）",
    responses={
        200: {
            "content": {"application/octet-stream": {}},
            "description": "附件文件内容",
        }
    },
)
async def download_attachment_by_id(
    attachment_id: UUID,
    service: AttachmentServiceDep,
) -> StreamingResponse:
    """按附件 ID 下载附件内容"""
    from app.repositories.attachment_repo import AttachmentRepository
    from app.config.minio_client import MinIOClient
    from app.utils.exceptions import NotFoundException

    repo = AttachmentRepository(service.session)
    attachment = await repo.get_by_id(attachment_id)
    if not attachment:
        from fastapi import HTTPException
        raise HTTPException(status_code=404, detail="附件不存在")

    content = MinIOClient.download_file(attachment.object_name)

    return StreamingResponse(
        io.BytesIO(content),
        media_type=attachment.content_type,
        headers={
            "Content-Disposition": f'attachment; filename="{attachment.file_name}"',
            "Content-Length": str(len(content)),
        }
    )
